File: java/MinStack.java
public class MinStack {
    private Stack<Integer> stack;
    private Stack<Integer> minStack;

    public Solution() {
        // do initialize if necessary
        stack = new Stack<Integer>();
        minStack = new Stack<Integer>();
    }

    public void push(int number) {
        // write your code here
        stack.push(number);
        if (minStack.isEmpty() || minStack.peek() >= number) {
            minStack.push(number);
        }
    }

    public int pop() {
        // write your code here
        int number = stack.pop();
        if (number <= minStack.peek()) {
            minStack.pop();
        }
        return number;
    }

    public int min() {
        // write your code here
        return minStack.peek();
    }
}

File: python/min_stack.py
class MinStack:
    def __init__(self):
        self.stack = []
        self.minstack = []

    def push(self, number):
        self.stack.append(number)
        if not self.minstack or self.minstack[-1] >= number:
            self.minstack.append(number)

    def pop(self):
        number = self.stack.pop()
        if number == self.minstack[-1]:
            self.minstack.pop()
            return number

    def min(self):
        return self.minstack[-1]
